Ccip 5847  - Convert DeployContracts into operation API based calls

AnieeG opened the pull request:
Wraps all deploy contract logic into individual operations.
Intent is to create separate ops file for each contract so that it's easier to locate the corresponding ops for each contract write operations.
As part of this moved all deployContracts, setRMNRemoteConfig to individual operation and sequence.
Made this folder struct -
```
   ccip
   | -- sequence
         |--- v1_6
   |-- operations 
         |-- evm 
              |-- v1_2
                  |-- ops_router
              |--v1_6
                  |-- ops_noncemanager and so on 
```

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into develop

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into develop

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into develop

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into develop

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into develop

AnieeG committed:
add ops

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into CCIP-5847-ops

AnieeG committed:
more

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into CCIP-5847-ops

AnieeG committed:
changes

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into CCIP-5847-ops

AnieeG committed:
changes

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into CCIP-5847-ops

AnieeG committed:
ops

AnieeG committed:
updates

AnieeG committed:
changes

AnieeG committed:
Merge branch 'develop' of github.com:smartcontractkit/chainlink into CCIP-5847-ops

AnieeG committed:
fixes

AnieeG committed:
lint fixes

graham-chainlink commented on deployment/ccip/changeset/v1_6/cs_deploy_chain_test.go:
```diff
@@ -121,9 +123,10 @@ func TestDeployChainContractsChangeset(t *testing.T) {
 
 	// try to deploy chain contracts again and it should not deploy any new contracts except feequoter
 	// but should not error
+	// creating a new environment to avoid reusing the old one so that the sequence gets executed again
```
this comment is not needed anymore right?

AnieeG committed:
more lint fix

graham-chainlink commented on deployment/ccip/sequence/evm/v1_6/seq_rmn.go:
```diff
@@ -0,0 +1,101 @@
+				if err != nil {
+					return opsutil.OpOutput{}, fmt.Errorf("failed to aggregate proposals: %w", err)
+				}
+				deps.Env.Logger.Infow("Generated proposal for RMNRemoteConfig", "chains", maps.Keys(input.RMNRemoteConfigs))
```
we should use the logger in the bundle eg `b.Logger`

AnieeG committed:
review comments

RodrigoAD commented on deployment/ccip/operations/evm/v1_6/ops_fee_quoter.go:
```diff
@@ -0,0 +1,188 @@
+			state := deps.CurrentState
+			chain := deps.Env.Chains[input.Chain]
+			ab := deps.AddressBook
+			chainState, chainExists := state.Chains[input.Chain]
```
The Operation responsibility should be simply deploying the contract and returning an address. If the contract exists, the check should happen before we run the operation (in the sequence), this makes the operation predictable.

AnieeG replied:
@RodrigoAD I think it would make the designing a lot simple and complete if we have this check inside an operation. I can see no use case where an operation should deploy contracts if it exists. If for some reason one misses to have this check from the caller function it will unnecessary deploy a contract ( in turn generating extra cost)

RodrigoAD replied:
Operations by definition should be as pure as possible, so they can be composed later on in Sequences. If the output depends on the input, it makes the operation unpredictable and hard to compose. 

The smart and complex logic should live in Sequences, that's what they are for, and they come with default functionality to handle retries and idempotency (which avoids this double deployment expense you mention)

graham-chainlink replied:
Hi @RodrigoAD  just in case you didn know, the retry for operation is now opt in instead of on by default. - [docs](https://github.com/smartcontractkit/chainlink-deployments-framework/pull/108)

This was updated few version back after a discussion with the team.

RodrigoAD commented on deployment/ccip/operations/evm/v1_6/ops_fee_quoter.go:
```diff
@@ -0,0 +1,188 @@
+
+			contractParams := input.Params
+			if chainState.FeeQuoter == nil {
+				feeQ, err := cldf.DeployContract(b.Logger, chain, ab,
```
I would discourage using inderterministic functions like `DeployContract` that perform multiple side effects inside operations, in this case sending a tx, waiting and writing to the address book. Multiple side effects should always be composed in Sequences

AnieeG replied:
@RodrigoAD If I change this now the scope of this PR will increase and the purpose of this PR is not to do any existing code changes but just to wrap the existing logic in Operation API. I can create a ticket for evaluating and implementing this later for all deployContractOps

graham-chainlink replied:
I do think there is value in what @RodrigoAD is saying, similar reasons to what i commented [here](https://github.com/smartcontractkit/chainlink/pull/17842/files#r2102916819)

RodrigoAD replied:
> If I change this now the scope of this PR will increase and the purpose of this PR is not to do any existing code changes but just to wrap the existing logic in Operation API

How the team organizes the work to migrate to Operations is up to the team of course, I'm suggesting we can do this in one effort following best practices, so you don't have to refactor your code twice, and we can start leveraging the work from the beginning in Aptos and Sui :)

RodrigoAD commented on deployment/ccip/operations/evm/v1_6/ops_offramp.go:
```diff
@@ -0,0 +1,113 @@
+		"Deploys OffRamp 1.6 contract on the specified evm chain",
+		func(b operations.Bundle, deps opsutil.OpDependencies, input DeployOffRampInput) (common.Address, error) {
+			state := deps.CurrentState
+			e := deps.Env
```
We must avoid passing the Environment down to Operations or we'll end up having the same issues as changesets. Operations should only require the dependencies they need

cc @graham-chainlink

AnieeG replied:
What issues did we face for changeset for passing Env?

graham-chainlink replied:
passing whole env as dependency just feels like an anti pattern , the bundle provide the logger and context so we dont need the environment for the standard stuff. 

if we pass what we need in the operations in the deps, then it is clearer to the users what the operation needs. For example the env contains fields such as Bundle which wont be needed.

if we have to pass the whole env down, then maybe the operation is doing too many things?

RodrigoAD replied:
As @graham-chainlink explains, is an anti-pattern. Operations should just require what they need, so they are easily consumable for the user. 
Passing the entire environment can lead to undesired effects, the CLD team faced some issues already that they mitigated with [Overrides](https://docs.cld.cldev.sh/guides/migrations/migration-environment-override#withoutjd)

AnieeG replied:
I just tried removing env but could not as we heavily use DeployerGroup in all cs and it needs env as parameter - https://github.com/smartcontractkit/chainlink/blob/8b399f5f60e883168eeed006eb86691bf214096f/deployment/ccip/shared/deployergroup/deployer_group.go#L151
I removed it for DeployContract related operation though, but need to retain it for now for all operation which are making config changes in contracts

RodrigoAD commented on deployment/ccip/operations/evm/v1_6/ops_offramp.go:
```diff
@@ -0,0 +1,113 @@
+				return common.Address{}, fmt.Errorf("chain %s not found in existing state, "+
+					"deploy the prerequisites first", chain.String())
+			}
+			if chainState.RMNProxy == nil {
```
All these params should come from the input `DeployOffRampInput` directly. If the information lives in the environment, let's prepare it beforehand (in the changeset, pulling the info from the environment) and call the operation/sequence then

RodrigoAD replied:
@graham-chainlink Do we have an example on how to structure Operations and Sequences like these?

graham-chainlink replied:
yeh @RodrigoAD , we are working on building a set of best practices for operations api. 

cc @jkongie more samples for our best practice doc :D

RodrigoAD reviewed (commented):
Hey @AnieeG, very happy to see Operations being leveraged :) 

I left some suggestions on good practices we should consider. In Aptos we'll soon need Operations and Sequences like these and I'd love to reuse them

kylesmartin commented on deployment/ccip/operations/evm/v1_2/ops_router.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+var (
+	DeployRouter = operations.NewOperation(
```
Let's limit the scope of what operations can do as much as possible. For example, for deployment and contract operations more generally our dependency should just be a chain client and we should return the new address / MCMS proposal. Environment management, such as saving addresses, should happen in a different changeset. We should align with the single side-effect rule because it helps the pipeline be as granular as possible when retrying operations in a sequence. In this case, we are saving to the environment and deploying a contract, which would be 2 side effects.

jkongie replied:
[Here is an operation](https://github.com/smartcontractkit/chainlink/blob/29543efb082f24212a5c953bc48f0e7be4ead03c/deployment/tokens/internal/ops/ops_storage.go#L11-L58) which adds an address to the address book. You may copy it into your package if you need.

We are considering adding common operations to the framework for reusability purposes, and you may see it there next week

AnieeG replied:
If we want to change this, I will have to change DeployContract logic, which will create a drastic change in the existing code. IMO it's better to do this incrementally or evaluate how useful is this to have this level of granularity. Will we ever deploy contract without saving it in addressbook? if we do, are these contracts part of this? I understand conforming to operation standard is important but in this case it will cost some significant changes to existing code . is this absolutely necessary?

graham-chainlink replied:
I do echo what @kylesmartin is saying, operation is meant to be granular and sequence gives that extra composing ability on top.

With the example of DeployContract and saving to AB, if we break it down to 2 operations
- 1 to deployContract
- 1 to saveAB

This gives benefit that it is a single side effect (which makes it granular too) so we can take advantage of the idempotency feature. 

Also if in future , we dont want to save to AB , but another datastore (like how currently we are moving from addressbook to datastore), we could just compose a new sequence which contains 1 operation to deployContract and 1 operation to save to Datastore.

Then we avoid modifying the operation DeployContract (just good separation) and the saving logic being in an operation itself can be reused else where too.

AnieeG replied:
Discussed with @kylesmartin on this - We would like to keep the current function for now considering there is no immediate value for us to introduce a massive code change related to this. However when the common ops for persist address is merged - we will come back to this and update the it accordingly.

kylesmartin commented on deployment/ccip/sequence/evm/v1_6/seq_rmn.go:
```diff
@@ -0,0 +1,101 @@
+)
+
+var (
+	SetRMNRemoteConfigSequence = operations.NewSequence(
```
Can we make sequences and operations internal to deployment/ccip?

My general thinking was that sequences orchestrate ops, and sequences are exposed 1:1 as changesets. So we should never need to expose operations or sequences to users. WDYT?

rodrigombsoares replied:
> and sequences are exposed 1:1 as changesets    

This is not true. A changeset may run N sequences and another chain family cs may need to call an EVM sequence for something.

AnieeG replied:
We should not make sequences and operations internal. It will limit us if we need to create dynamic changeset within migration framework in CLD on-demand basis.

rodrigombsoares commented on deployment/ccip/operations/evm/v1_6/ops_fee_quoter.go:
```diff
@@ -0,0 +1,188 @@
+		"DeployFeeQuoter",
+		semver.MustParse("1.0.0"),
+		"Deploys FeeQuoter 1.6 contract on the specified evm chain",
+		func(b operations.Bundle, deps opsutil.OpDependencies, input DeployFeeQInput) (common.Address, error) {
```
Can we use regular functions instead of anonymous?
It will remove nested code and improve readability and maintainability.

rodrigombsoares commented on deployment/ccip/sequence/evm/v1_6/seq_deploy_chain.go:
```diff
@@ -0,0 +1,362 @@
+	return nil
+}
+
+var (
```
We could unnest this code. There's no reason to use the group syntax.

```
var DeployChainContractsSeq = operations.NewSequence(
    ...
)
```

rodrigombsoares commented:
very nit, but if we name `changeset` and `sequence` it would be nice if `operations` would follow as `operation`

graham-chainlink reviewed (commented):
nitpick: i think renaming the ccip/operations folder to ccip/operation so it is singular instead of plural, aligning with go convention, similar to how we have sequence (singular)

Update: looks like me and @RodrigoAD  got the same comment :D

AnieeG committed:
input the addresses instead of reading from state

AnieeG committed:
rename

AnieeG committed:
review comments

AnieeG commented:
@RodrigoAD @kylesmartin @jkongie @rodrigombsoares @graham-chainlink - Please review it again - Based on immediate need and the existing code related constraints, I have applied the following changes -
1. Created a subset of dependency for DeployContract related ops - removed env, state.
2. Renamed `operations` to `operation`
3. Updated DeployContract ops to always deploy contracts and moved the check for existing contracts in sequence.
4. Used parameters to pass addresses to operation instead of reading from state

Things that we can do in future -
1. Move persist Address out of deployContract operation when the common operation is available.
2. Try to evaluate for Configure contract related ops if environment can be removed as dependency. currently it's needed for using deployergroup in operation - example - https://github.com/smartcontractkit/chainlink/blob/CCIP-5847-ops/deployment/ccip/operation/evm/v1_6/ops_rmnremote.go#L102-L111

AnieeG committed:
fix error
